### ISSUE TITLE
[dargon2] Failed to load dynamic lib - macOS M1

### ISSUE THREAD

**Issue by jayjah** [tmthecoder/dargon2#7]
Code example from your Readme:
```
Future<String> generatePassword3() async {
  var password = 'password';
  //use Salt(List<int> bytes) for a salt from an Integer list
  var s = Salt.newSalt();
  print('Generating password hash');
  //Hash with pre-set params (iterations: 32, memory: 256, parallelism: 2,
  //length: 32, type: Argon2Type.i, version: Argon2Version.V13)
  var result = await argon2.hashPasswordString(
    password,
    salt: s,
    type: Argon2Type.id,
  );
  //  memory: 131072,
  //  iterations: 8,
  // parallelism: 4);

  print('Result: $result');
  //Raw hash values available as int list, base 64 string, and hex string
  //var bytesRaw = result.rawBytes;
  //var base64Hash = result.base64String;
  //var hexHash = result.hexString;

  //Encoded hash values available as int list and encoded string
  //var bytesEncoded = result.encodedBytes;
  var stringEncoded = result.encodedString;
  print('Result as string: $stringEncoded');

  //Verify password (returns true/false), uses default type (Argon2Type.i)
  var verified = await argon2.verifyHashString(password, stringEncoded);
  print(
      'Password: $password :: encdeod data: $stringEncoded :: Match: $verified');

  return stringEncoded;
}
```

Crash reported:
```
$ dart bin/argon_test.dart
Generating password hash

===== CRASH =====
si_signo=Illegal instruction: 4(4), si_code=1, si_addr=0x11345306b
version=2.16.1 (stable) (Tue Feb 8 12:02:33 2022 +0100) on "macos_x64"
pid=68319, thread=21507, isolate_group=main(0x7f7bf300d000), isolate=main(0x7f7bf3012000)
isolate_instructions=1025db340, vm_instructions=1025db340
  pc 0x000000011345306b fp 0x000000030ba55300 argon2_hash+0xeb
  pc 0x000000010e48598b fp 0x000000030ba55358 Unknown symbol
  pc 0x0000000113c6cd2e fp 0x000000030ba553e8 Unknown symbol
  pc 0x0000000113c669c3 fp 0x000000030ba555a0 Unknown symbol
  pc 0x0000000113c65df7 fp 0x000000030ba55670 Unknown symbol
  pc 0x0000000113c65afc fp 0x000000030ba55758 Unknown symbol
  pc 0x0000000113c649e1 fp 0x000000030ba55828 Unknown symbol
  pc 0x0000000113c24d48 fp 0x000000030ba558c8 Unknown symbol
  pc 0x0000000113c248fe fp 0x000000030ba558f8 Unknown symbol
  pc 0x0000000113c24391 fp 0x000000030ba55978 Unknown symbol
  pc 0x0000000113c237e1 fp 0x000000030ba559a8 Unknown symbol
  pc 0x0000000113c2344c fp 0x000000030ba559d8 Unknown symbol
  pc 0x0000000113c2264c fp 0x000000030ba55a40 Unknown symbol
  pc 0x0000000113c2288d fp 0x000000030ba55a78 Unknown symbol
  pc 0x0000000113c2264c fp 0x000000030ba55ae0 Unknown symbol
  pc 0x0000000113c21ec3 fp 0x000000030ba55b30 Unknown symbol
  pc 0x000000010e4829bf fp 0x000000030ba55ba8 Unknown symbol
  pc 0x0000000102747bed fp 0x000000030ba55c50 dart::DartEntry::InvokeCode(dart::Code const&, unsigned long, dart::Array const&, dart::Array const&, dart::Thread*)+0x14d
  pc 0x0000000102747a28 fp 0x000000030ba55cb0 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)+0x128
  pc 0x0000000102749d15 fp 0x000000030ba55d00 dart::DartLibraryCalls::HandleMessage(long long, dart::Instance const&)+0x155
  pc 0x00000001027716cb fp 0x000000030ba55de0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)+0x35b
  pc 0x000000010279b67c fp 0x000000030ba55e50 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x12c
  pc 0x000000010279bd8f fp 0x000000030ba55ea0 dart::MessageHandler::TaskCallback()+0x1df
  pc 0x00000001028ca6a8 fp 0x000000030ba55f30 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x138
  pc 0x00000001028cab3d fp 0x000000030ba55f60 dart::ThreadPool::Worker::Main(unsigned long)+0x5d
  pc 0x000000010283b13f fp 0x000000030ba55fb0 dart::OSThread::GetMaxStackSize()+0xaf
  pc 0x00007ff80afa14f4 fp 0x000000030ba55fd0 _pthread_start+0x7d
  pc 0x00007ff80af9d00f fp 0x000000030ba55ff0 thread_start+0xf
-- End of DumpStackTrace
  pc 0x0000000000000000 fp 0x000000030ba55358 sp 0x0000000000000000 [Optimized] 2266775552
  pc 0x0000000113c6cd2e fp 0x000000030ba553e8 sp 0x000000030ba55368 [Optimized] FfiTrampoline
  pc 0x0000000113c669c3 fp 0x000000030ba555a0 sp 0x000000030ba553f8 [Unoptimized] DArgon2Native.hashPasswordBytesSync
  pc 0x0000000113c65df7 fp 0x000000030ba55670 sp 0x000000030ba555b0 [Unoptimized] DArgon2Native.hashPasswordBytes
  pc 0x0000000113c65afc fp 0x000000030ba55758 sp 0x000000030ba55680 [Unoptimized] DArgon2Native.hashPasswordBytes
  pc 0x0000000113c649e1 fp 0x000000030ba55828 sp 0x000000030ba55768 [Unoptimized] DArgon2.hashPasswordString
  pc 0x0000000113c24d48 fp 0x000000030ba558c8 sp 0x000000030ba55838 [Unoptimized] generatePassword3
  pc 0x0000000113c248fe fp 0x000000030ba558f8 sp 0x000000030ba558d8 [Unoptimized] generatePassword3
  pc 0x0000000113c24391 fp 0x000000030ba55978 sp 0x000000030ba55908 [Unoptimized] main
  pc 0x0000000113c237e1 fp 0x000000030ba559a8 sp 0x000000030ba55988 [Unoptimized] main
  pc 0x0000000113c2344c fp 0x000000030ba559d8 sp 0x000000030ba559b8 [Unoptimized] main
  pc 0x0000000113c2264c fp 0x000000030ba55a40 sp 0x000000030ba559e8 [Unoptimized] _Closure@0150898.dyn:call
  pc 0x0000000113c2288d fp 0x000000030ba55a78 sp 0x000000030ba55a50 [Unoptimized] _delayEntrypointInvocation@1026248.<anonymous closure>
  pc 0x0000000113c2264c fp 0x000000030ba55ae0 sp 0x000000030ba55a88 [Unoptimized] _Closure@0150898.dyn:call
  pc 0x0000000113c21ec3 fp 0x000000030ba55b30 sp 0x000000030ba55af0 [Unoptimized] _RawReceivePortImpl@1026248._handleMessage@1026248
  pc 0x000000010e4829bf fp 0x000000030ba55ba8 sp 0x000000030ba55b40 [Stub] InvokeDartCode
[1]    68319 abort      dart bin/argon_test.dart
```

Version:
```
Dart SDK version: 2.16.1 (stable) (Tue Feb 8 12:02:33 2022 +0100) on "macos_x64"
```

**Comment by tmthecoder** [comment]
@jayjah I'm on a macbook pro (intel) and I ran the same code you gave me. Aside from the error where the `argon2.verifyHashString` fails as the `type` parameter isn't set, I don't see any errors.

Would you be able to link your entire source project so I could try to clone/build it? 

Here's the code I used which uses `Argon2Type.id` in the verify call:

```dart
Future<String> generatePassword3() async {
  var password = 'password';
  //use Salt(List<int> bytes) for a salt from an Integer list
  var s = Salt.newSalt();
  print('Generating password hash');
  //Hash with pre-set params (iterations: 32, memory: 256, parallelism: 2,
  //length: 32, type: Argon2Type.i, version: Argon2Version.V13)
  var result = await argon2.hashPasswordString(
    password,
    salt: s,
    type: Argon2Type.id,
  );
  //  memory: 131072,
  //  iterations: 8,
  // parallelism: 4);

  print('Result: $result');
  //Raw hash values available as int list, base 64 string, and hex string
  //var bytesRaw = result.rawBytes;
  //var base64Hash = result.base64String;
  //var hexHash = result.hexString;

  //Encoded hash values available as int list and encoded string
  //var bytesEncoded = result.encodedBytes;
  var stringEncoded = result.encodedString;
  print('Result as string: $stringEncoded');

  //Verify password (returns true/false), uses default type (Argon2Type.i)
  var verified = await argon2.verifyHashString(password, stringEncoded, type: Argon2Type.id);
  print(
      'Password: $password :: encdeod data: $stringEncoded :: Match: $verified');

  return stringEncoded;
}
```

Here's the result I get:

```
Generating password hash
Result: Instance of 'DArgon2Result'
Result as string: $argon2id$v=19$m=256,t=32,p=2$sD/iZTXcmsKkAW4n7FdYQA$ZhIRFM38PJDeTHnAaHymVO1tsAW8imLqVRTCwYGUJxc
Password: password :: encdeod data: $argon2id$v=19$m=256,t=32,p=2$sD/iZTXcmsKkAW4n7FdYQA$ZhIRFM38PJDeTHnAaHymVO1tsAW8imLqVRTCwYGUJxc :: Match: true

Process finished with exit code 0
```

**Comment by jayjah** [comment]
@tmthecoder Of course, that's not the problem. But I think I found the problem by looking at your answer and my dart version. Actually I am running this on an Apple MacBook Air M1. I just downloaded the correct Dart SDK, cleaned my pub-cache und ran it again. Now it seems to be clear, why it's failing.
Especially because you are running this from an Intel CPU. 

```
./dart-sdk/bin/dart run bin/argon_test.dart 
Generating password hash
Unhandled exception:
Invalid argument(s): Failed to load dynamic library '/Users/jayjah/.pub-cache/hosted/pub.dartlang.org/dargon2-3.1.0/lib/src/blobs/libargon2-darwin.dylib': dlopen(/Users/jayjah/.pub-cache/hosted/pub.dartlang.org/dargon2-3.1.0/lib/src/blobs/libargon2-darwin.dylib, 0x0001): tried: '/Users/jayjah/.pub-cache/hosted/pub.dartlang.org/dargon2-3.1.0/lib/src/blobs/libargon2-darwin.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/libargon2-darwin.dylib' (no such file)
#0      _open (dart:ffi-patch/ffi_dynamic_library_patch.dart:12:43)
#1      new DynamicLibrary.open (dart:ffi-patch/ffi_dynamic_library_patch.dart:23:12)
#2      DartLibLoader.loadLib (package:dargon2/src/native/dart_lib_loader.dart:23:27)
#3      new LocalBinder._ (package:dargon2_core/src/native/local_binder.dart:155:31)
#4      LocalBinder.initialize (package:dargon2_core/src/native/local_binder.dart:130:38)
#5      new DArgon2Native (package:dargon2_core/src/native/dargon2_native.dart:19:17)
#6      argon2 (package:dargon2/src/argon2.dart:11:16)
#7      argon2 (package:dargon2/src/argon2.dart)
#8      generatePassword3 (package:argon_test/argon_test.dart:95:22)
#9      main (file:///Users/jayjah/github/argon_test/bin/argon_test.dart:4:42)
#10     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
#11     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
```

I am using asdf to keep my languages and corresponding versions up to date, which silently ignores arm64 architecture. But that's another story. 

Now my dart version is the following: 
```
Dart SDK version: 2.16.1 (stable) (Tue Feb 8 12:02:33 2022 +0100) on "macos_arm64"
```

**Comment by tmthecoder** [comment]
@jayjah I think I know why it downloads the intel version. My release of the underlying argon2 C library has remained the same since pre-M1 times, so there isn't an arm (or universal) build available. 

Thanks for bringing this up, it completely slipped my mind. 

I've added the builder to create a universal binary now. Would you mind testing it as I don't have an M1 Mac?

Set your package declaration to:

```
dargon2:
    git:
        url: https://github.com/tmthecoder/dargon2.git
        path: dargon2
```

**Comment by jayjah** [comment]
That's awesome! Thanks for your fast feedback! :-)
Yes, I'll check it tomorrow, as soon as possible. Afterwards I'll keep you up to date!

**Comment by jayjah** [comment]
@tmthecoder Again thank you for your help!
The branch you mentioned worked as it should.

Now the output is the following:
```
$ dart run bin/argon_test.dart 
Generating password hash
Result: Instance of 'DArgon2Result'
Result as string: $argon2id$v=19$m=256,t=32,p=2$0nrqWuqhy8cr+ukZ+DPeKg$ku1oOW7WsvPqdEB/s96dSToR9q56qNfTvUf94k6FwwE
Password: password :: encdeod data: $argon2id$v=19$m=256,t=32,p=2$0nrqWuqhy8cr+ukZ+DPeKg$ku1oOW7WsvPqdEB/s96dSToR9q56qNfTvUf94k6FwwE :: Match: true
```

Please close this Issue as soon as you released the new version.

**Comment by tmthecoder** [comment]
Thanks for testing!

The changes are available on v3.2.0
